src/marqo/core/typed_vespa_index.py

```diff
@@ -0,0 +1,271 @@
+        Generate the document (fields) section of the Vespa schema. Update `marqo_index` with Vespa-level field names.
+        """
+        schema.append(f'document {marqo_index.name} {{')
+        for field in marqo_index.fields:
+            field_type = cls._get_vespa_type(field.type)
+
+            if FieldFeature.LexicalSearch in field.features:
+                field_name = f'{cls._INDEX_FIELD_PREFIX}{field.name}'
+                schema.append(f'field {field_name} type {field_type} {{')
+                schema.append(f'indexing: index | summary')
+                schema.append('index: enable-bm25')
+                schema.append('}')
+
+                field.lexical_field_name = field_name
+
+            if FieldFeature.Filter in field.features:
+                field_name = f'{cls._FILTER_FIELD_PREFIX}{field.name}'
+                schema.append(f'field {field_name} type {field_type} {{')
+                schema.append('indexing: attribute | summary')
+                schema.append('attribute: fast-search')
+                schema.append('rank: filter')
+                schema.append('}')
+
+                field.filter_field_name = field_name
+
+            if FieldFeature.LexicalSearch not in field.features and FieldFeature.Filter not in field.features:
+                field_name = field.name
+                schema.append(f'field {field_name} type {field_type} {{')
+                schema.append('indexing: summary')
+                schema.append('}')
+
+        # score modifiers
+        if marqo_index.score_modifier_fields:
+            schema.append(f'field {cls._SCORE_MODIFIERS_FIELD} type tensor<float>(p{{}}) {{ indexing: attribute }}')
+
+        # tensor fields
+        model_dim = cls._get_model_dimension(marqo_index)
+        for field in marqo_index.tensor_fields:
+            chunks_field_name = f'{cls._CHUNKS_FIELD_PREFIX}{field.name}'
+            embedding_field_name = f'{cls._EMBEDDING_FIELD_PREFIX}{field.name}'
+            schema.append(f'field {chunks_field_name} type array<string> {{')
+            schema.append('indexing: attribute | summary')
+            schema.append('}')
+            schema.append(f'field {embedding_field_name} type tensor<float>(p{{}}, x[{model_dim}]) {{')
+            schema.append('indexing: attribute | index | summary')
+            schema.append(f'attribute {{ distance-metric: {cls._get_distance_metric(marqo_index.distance_metric)} }}')
+            schema.append('index { hnsw {')
+            schema.append(f'max-links-per-node: {marqo_index.hnsw_config.m}')
+            schema.append(f'neighbors-to-explore-at-insert: {marqo_index.hnsw_config.ef_construction}')
+            schema.append('}}')
+            schema.append('}')
+
+            field.chunk_field_name = chunks_field_name
+            field.embeddings_field_name = embedding_field_name
+
+        schema.append('}')
+
+    @classmethod
+    def _generate_summaries(cls, marqo_index: MarqoIndex, schema: List[str]) -> None:
```

Comment:
Is it possible to keep this method as pure as possible? So not mutating `schema`, and instead returning a new one? I am not a huge fan of these mutating-style methods

Reply:
The alternative would be to clone that list, a linear time operation every time (linear in total length of all strings combined).